src/classes/ADDR_Contact_TDTM.cls

```diff
@@ -188,9 +200,8 @@ public class ADDR_Contact_TDTM extends TDTM_Runnable {
                     continue;
                 }
 
-                // if the address changed, remember the contact we want to add a new address for    
-                Contact conOld = (listOld != null ? (Contact)listOld[i] : null);
-                if (isContactAddressChanged(con, conOld)) 
+                // if the contacts address changed or if the contact with an address moves to a new household remember the contact
+                if (isContactAddressChanged(con, conOld) || (acctChanged && !isContactAddressEmpty(con))) 
```

Comment:
and I take the blame that this code is confusing!  I wrote it, and it was pretty hard to figure out the best way to structure the code and make it follow the precedence rules I came up with and deal with all the permutations of potential changes!  and of course there was no official spec!